### PR TITLE
Show text search bar when the filter is collapsed

### PR DESCRIPTION
**Pull request by syncguru**
Fixes #20102

Filter box collapsed:
screen shot 2016-08-21 at 3 34 37 pm

Filter box expanded:
screen shot 2016-08-21 at 3 34 50 pm

### CODE REVIEW

**Comment by vboctor** [comment]
It would be great if we can have "Search" next to the text box and also provide a way to select from the saved filters.  Here is the collapsed version from the classic UI.

filter_collapsed

**Comment by syncguru** [comment]
@vboctor the classic view is way too busy for a collapsed filter. I've a simpler collapsed filter bar for modern ui:

screen shot 2016-08-22 at 9 06 25 pm

screen shot 2016-08-22 at 9 18 43 pm

**Comment by atrol** [comment]
While you are on it, some more ideas:

- Add a "Reset Filter" button
- Remove the [Reset Filter] entry from drop down list
- Remove the "Use Filter" button (no longer needed as we agreed some while ago that JavaScript must be enabled for MUI)

"Save Current Filter" and "Create Permalink" are similar operations.
It's a bit confusing to have one of them as a button and the other one as a menu entry.
What about removing the buttons "Save Current Filter" and "Manage Filters" and add them to the menu?

**Comment by dregad** [comment]
:+1: with atrol's suggestions

**Comment by vboctor** [comment]
In the case where the user has put together an adhoc filter, what would be shown in the filter name dropdown?  Suggestions to avoid confusion?

**Comment by dregad** [comment]
`Unsaved filter` maybe ?

**Comment by syncguru** [comment]
@atrol I did try to work through your suggestions, here is what I ended up with:
1- "Use Filter" button is gone. Now, form submission is triggered on dropdown change
2-  Removing [Reset Filter] option and switching that to a button felt like undoing (1) .. having no button to control filter selection is much better
3- Moved the 'manage filter' to menu
4- Left "Save current Filter" as is for couple of reasons:
a. In the menu, saving was hard to find.
b. Saving filter requires changes to filter JS logic which should be its own issue/PR. I think (a) was really bad so (b) is not very relevant.

screen shot 2016-08-25 at 7 19 28 pm

screen shot 2016-08-25 at 7 19 46 pm

**Comment by vboctor** [comment]
:+1:

**Comment by atrol** [comment]
Issue 1:
Try collapsed filter without having a stored filter.
Resetting filter is not that intuitive.
You have to select the empty row in the drop down and change to "Reset Filter" after that.

Issue 2:
Try collapsed filter with one or more stored filters.
Select a filter from drop down.
The drop down does not show the chosen filter after that.

Issue 3:
The "Search" in text input field is not complete visible in collapsed mode.
Seems to be just a Safari issue.
searchcollapsed

Issue 4:
Strings entered in the search field are not displayed after applying the filter.
You get "Search" again.

**Comment by syncguru** [comment]
Rebased and fixed #3 & #4 in @atrol list

**Comment by atrol** [comment]
There are changes in the PR (verify.php) that are not related to search bar.

**Comment by syncguru** [comment]
@atrol Fixed the PR. As a quick recap of your list of issues:
#1 dropdown is hidden when there are no stored filters
#2 fixed
#3 fixed
#4 fixed

File: core/filter_api.php [hunk 1]
```diff
@@ -2315,31 +2316,76 @@ function filter_draw_selection_area2( $p_page_number, $p_for_screen = true, $p_e
...
-					}
-				}
-				?>
-				<?php
-				if( access_has_project_level( config_get( 'create_permalink_threshold' ) ) ) {
-					echo '<a class="btn btn-xs btn-white btn-primary btn-round" href="permalink_page.php?url=', urlencode( filter_get_url( $t_filter ) ), '">';
-					echo lang_get( 'create_filter_link' );
-					echo '</a>';
-				} ?>
+			<?php if ( count( $t_stored_queries_arr ) > 0 ) { ?>
```
Comment (atrol) [inline core/filter_api.php]: There should be no blank between if and (

File: core/filter_api.php [hunk 2]
```diff
@@ -3613,24 +3659,16 @@ function filter_draw_selection_area2( $p_page_number, $p_for_screen = true, $p_e
 	echo '<div class="btn-toolbar pull-right">';
 	echo '<div class="btn-group">';
 
-	$t_stored_queries_arr = filter_db_get_available_queries();
-
-	if( access_has_project_level( config_get( 'stored_query_create_threshold' ) ) ) { ?>
+	if ( access_has_project_level( config_get( 'stored_query_create_threshold' ) ) ) { ?>
```
Comment (atrol) [inline core/filter_api.php]: Same as above

File: core/filter_api.php [hunk 3]
```diff
@@ -3613,24 +3659,16 @@ function filter_draw_selection_area2( $p_page_number, $p_for_screen = true, $p_e
...
 	}
-	if( count( $t_stored_queries_arr ) > 0 ) { ?>
-
-	<form class="form-inline pull-left" method="post" name="open_queries" action="query_view_page.php">
-			<?php # CSRF protection not required here - form does not result in modifications ?>
-			<input type="submit" name="switch_to_query_button" class="btn btn-primary btn-white btn-sm btn-round"
-				value="<?php echo lang_get( 'open_queries' )?>" />
-		</form>
-		<form class="form-inline pull-left padding-left-8"  method="get" name="list_queries<?php echo $t_form_name_suffix;?>" action="view_all_set.php">
+	if ( count( $t_stored_queries_arr ) > 0 ) { ?>
```
Comment (atrol) [inline core/filter_api.php]: Same as above

**Comment by atrol** [comment]
Thanks for fixing the issues.
I tested a bit and did not find any obvious bug.

What I still don't like that much is the way to reset the filter if there exist stored filters.
To be sure this is not caused by my testing environment:
After changing the filter you see this
resetfilter1

After opening the menu you get 
resetfilter2


You have to select the empty entry to reset the filter.
Is that what you expected?
If so, I think it's not that intuitive. 

If you still don't like the additional button to reset the filter:
What about displaying [Unsaved Filter] instead of the empty line and selecting this entry
instead of [Reset Filter]?
So you would open the drop down, choose [Reset Filter] to reset and will get [Unsaved Filter]
again.

**Comment by syncguru** [comment]
@atrol - I checked a really simple fix - move [Reset Filter] to the second spot in the dropdown list instead of first. I think this should deliver the expected behavior. Give it a try and let me know.

I prefer not having an additional button for this. MantisBT has been around for a long time and if this was real pain point, it would be been fixed long time ago. I'd like to not optimize for the rare usage.

**Comment by atrol** [comment]
> move [Reset Filter] to the second spot in the dropdown list 

It's better than before, but I still don't like the UI that much (same in 1.3.x).
It does not mean that I want to block this PR as the main reason for it is to show the
text search when the filter is collapsed.
Thus, +1 after having applied my minor requests for the white space issues.

**Comment by vboctor** [comment]
I did some testing and it looks good.  I would just replace the blank element in the dropdown with '[select filter]', but we can do this as a follow up

**Pull request merged**